test/system/src/test/java/io/pravega/test/system/MetadataScalabilityTest.java

```diff
@@ -0,0 +1,179 @@
+        ControllerImpl controller = getController();
+
+        // manually scale the stream SCALES_TO_PERFORM times
+        Stream stream = new StreamImpl(SCOPE, STREAM_NAME);
+        AtomicInteger counter = new AtomicInteger(0);
+        List<List<Segment>> listOfEpochs = new LinkedList<>();
+
+        CompletableFuture<Void> scaleFuture = Futures.loop(() -> counter.incrementAndGet() <= SCALES_TO_PERFORM,
+                () -> controller.getCurrentSegments(SCOPE, STREAM_NAME)
+                                .thenCompose(segments -> {
+                                    Map<Double, Double> newRanges = new HashMap<>();
+                                    double delta = 1.0 / NUM_SEGMENTS;
+                                    newRanges.put(delta * (counter.get() - 1), delta * counter.get());
+
+                                    ArrayList<Segment> sorted = Lists.newArrayList(segments.getSegments().stream()
+                                                                                           .sorted(Comparator.comparingInt(x ->
+                                                                                                   StreamSegmentNameUtils.getSegmentNumber(x.getSegmentId()) % NUM_SEGMENTS))
+                                                                                           .collect(Collectors.toList()));
+                                    log.info("found segments in epoch = {}", sorted);
+                                    listOfEpochs.add(sorted);
+                                    // note: with SCALES_TO_PERFORM < NUM_SEGMENTS, we can use the segment number as the index
+                                    // into the range map
+                                    List<Long> segmentsToSeal = sorted.stream()
+                                                                      .filter(x -> counter.get() - 1 == StreamSegmentNameUtils.getSegmentNumber(x.getSegmentId()) % NUM_SEGMENTS)
+                                                                      .map(Segment::getSegmentId).collect(Collectors.toList());
+                                    return controller.scaleStream(stream, segmentsToSeal, newRanges, executorService)
+                                                     .getFuture()
+                                                     .thenAccept(scaleStatus -> {
+                                                         log.info("scale stream for epoch {} completed with status {}", counter.get(), scaleStatus);
+                                                         assert scaleStatus;
+                                                     });
+                                }), executorService);
+
+        CompletableFuture<Void> result = scaleFuture
+                .thenCompose(r -> {
+                    // try SCALES_TO_PERFORM randomly generated stream cuts and truncate stream at those 
+                    // stream cuts. 
+                    List<AtomicInteger> indexes = new LinkedList<>();
```

Comment:
Why do the elements of this list need to be atomic integers? we only initialize them to `1` and never change after that, right? I'm not seeing the danger of concurrent access, but perhaps I'm missing something.

Reply:
actually this could have been an integer too.. 
`indexes` list is final anyway which is used in the future's callback (Futures.loop) which requires any variable used to be effectively final. 
The reason i used atomic integer though is because we take this `indexes` list and update values in this list in each iteration and that forms the next stream cut.. So alternate way of doing it would have been to take a fixed sized array and replace the previous value in the array with new value. that would have produced the same effect without having to use atomic integer

Reply:
You're right, you do get a reference and update.